File: ios-native/KPKNFit/KPKNFit/Data/Repository/CustomExerciseRepository.swift
import Foundation

public final class CustomExerciseRepository {

    static let shared = CustomExerciseRepository()

    private let db = KpknDatabase.instance()

    private var _customExercises: [ExerciseMuscleInfo] = []
    private let exercisesLock = NSLock()
    private var initialized = false

    public var customExercises: [ExerciseMuscleInfo] {
        exercisesLock.lock()
        defer { exercisesLock.unlock() }
        return _customExercises
    }

    private init() {}

    public static func getInstance() -> CustomExerciseRepository { shared }

    public func initialize() {
        if initialized { return }
        initialized = true
        Task {
            let entities = (try? await self.db.customExerciseDao.getAll()) ?? []
            let decoded = entities.compactMap { entity -> ExerciseMuscleInfo? in
                try? JSONDecoder().decode(ExerciseMuscleInfo.self, from: Data(entity.data.utf8))
            }
            let withCustomFlag = decoded.map { $0.copy(isCustom: true) }
            exercisesLock.lock()
            _customExercises = withCustomFlag
            exercisesLock.unlock()
            setCustomExerciseOverlay(exercises: withCustomFlag)
        }
    }

    public func upsert(exercise: ExerciseMuscleInfo) {
        if !initialized { return }
        let normalized = exercise.copy(isCustom: true)
        exercisesLock.lock()
        let filtered = _customExercises.filter { !$0.id.equalsIgnoreCase(normalized.id) }
        let sorted = (filtered + [normalized]).sorted { $0.name.lowercased() < $1.name.lowercased() }
        _customExercises = sorted
        exercisesLock.unlock()

        upsertCustomExerciseOverlay(exercise: normalized)
        Task {
            let data = try? JSONEncoder().encode(normalized)
            let entity = CustomExerciseEntity(
                id: normalized.id,
                name: normalized.name,
                data: data.map { String(decoding: $0, as: UTF8.self) } ?? "{}",
                createdAt: IsoDateFormatter.nowString(),
                updatedAt: IsoDateFormatter.nowString()
            )
            try? await self.db.customExerciseDao.upsert(entity: entity)
        }
    }

    public func delete(exerciseId: String) {
        if !initialized { return }
        exercisesLock.lock()
        _customExercises = _customExercises.filter { !$0.id.equalsIgnoreCase(exerciseId) }
        exercisesLock.unlock()

        removeCustomExerciseOverlay(exerciseId: exerciseId)
        Task {
            try? await self.db.customExerciseDao.delete(id: exerciseId)
        }
    }
}

private struct IsoDateFormatter {
    static func nowString() -> String {
        let formatter = ISO8601DateFormatter()
        return formatter.string(from: Date())
    }
}

private extension String {
    func equalsIgnoreCase(_ other: String) -> Bool {
        self.lowercased() == other.lowercased()
    }
}

File: android-native/app/src/main/java/com/example/kpkn/data/repository/CustomExerciseRepository.kt
package com.example.kpkn.data.repository

import android.content.Context
import com.example.kpkn.data.exercises.removeCustomExerciseOverlay
import com.example.kpkn.data.exercises.setCustomExerciseOverlay
import com.example.kpkn.data.exercises.upsertCustomExerciseOverlay
import com.example.kpkn.data.db.KpknDatabase
import com.example.kpkn.data.db.toEntity
import com.example.kpkn.data.db.toExerciseMuscleInfo
import com.example.kpkn.data.models.ExerciseMuscleInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

object CustomExerciseRepository {

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val _customExercises = MutableStateFlow<List<ExerciseMuscleInfo>>(emptyList())
    val customExercises: StateFlow<List<ExerciseMuscleInfo>> = _customExercises.asStateFlow()

    @Volatile
    private var initialized = false
    private lateinit var db: KpknDatabase

    fun initialize(context: Context) {
        if (initialized) return
        db = KpknDatabase.getInstance(context.applicationContext)
        initialized = true
        scope.launch {
            val saved = runCatching { db.customExerciseDao().getAll().map { it.toExerciseMuscleInfo().copy(isCustom = true) } }
                .getOrDefault(emptyList())
            _customExercises.value = saved
            setCustomExerciseOverlay(saved)
        }
    }

    fun upsert(exercise: ExerciseMuscleInfo) {
        if (!initialized) return
        val normalized = exercise.copy(isCustom = true)
        _customExercises.update { list ->
            val filtered = list.filterNot { it.id.equals(normalized.id, ignoreCase = true) }
            (filtered + normalized).sortedBy { it.name.lowercase() }
        }
        upsertCustomExerciseOverlay(normalized)
        scope.launch {
            runCatching {
                val existing = db.customExerciseDao().getById(normalized.id)
                db.customExerciseDao().upsert(normalized.toEntity(createdAt = existing?.createdAt))
            }
        }
    }

    fun delete(exerciseId: String) {
        if (!initialized) return
        _customExercises.update { list -> list.filterNot { it.id.equals(exerciseId, ignoreCase = true) } }
        removeCustomExerciseOverlay(exerciseId)
        scope.launch {
            runCatching { db.customExerciseDao().delete(exerciseId) }
        }
    }
}
